internal/pkg/EAR/heap.go

```diff
@@ -0,0 +1,154 @@
+package EAR
+
+import (
+	"go/types"
+	"golang.org/x/tools/go/ssa"
+	"strings"
+)
+
+// An abstract reference is a local (register) or global variable, possibly
+// annotated with some kind of context (call stack, allocation site etc).
+
+// AbsReference is the base class for local and global AbsReferences.
+type AbsReference interface {
+	DataType() types.Type // Return the data type of the underlying local or global
+	String() string
+}
+
+// AbsReferenceContext contains a stack of call instructions.
+type AbsReferenceContext []ssa.Instruction
+
+// AbsReferenceLocal is a heap partition that represents all abstract
+// objects that a register of reference type can point to in a specific context.
+type AbsReferenceLocal struct {
+	context *AbsReferenceContext
+	reg     ssa.Value
+}
+
+func (ref AbsReferenceLocal) DataType() types.Type {
+	return ref.reg.Type()
+}
+
+func (ref AbsReferenceLocal) String() string {
+	return ref.context.String() + ref.reg.Name()
+}
+
+// AbsReferenceGlobal is a heap partition that represents all abstract
+// objects that a global of reference type can point to.
+type AbsReferenceGlobal struct {
+	global *ssa.Global
+}
+
+func (ref AbsReferenceGlobal) DataType() types.Type {
+	return ref.global.Type()
+}
+
+func (ref AbsReferenceGlobal) String() string {
+	return ref.global.String()
+}
+
+// AbsReferenceInternal is for references created internally with an operator
+// and an operand reference. It is for creating references not directly associated
+// with an IR element. For example,
+// (1) *r0 can be represented by using operator VALUEOF (*);
+// (2) a unknown field can be represented by using operator FIELD.
+type AbsReferenceInternal struct {
```

Comment:
This is called ...Internal. When will users need to see this? How will they interact with it?

Reply:
As the comment specifies, it is supposed to be used by an analysis to create references that are not corresponding to the SSA elements. The SSA doesn't contain an element (e.g. a register, a global variable, or a parameter), but the heap needs to use it. So being "internal" means being internal to the analysis/heap. A better name may be used though.